testing/cuda/universal_pointer.cu

```diff
@@ -0,0 +1,32 @@
+#include <unittest/unittest.h>
+
+#include <thrust/allocate_unique.h>
+
+#include <vector>
+
+namespace
+{
+
+template <typename T>
+using universal_allocator = thrust::mr::stateless_resource_allocator<
+  T,
+  thrust::system::cuda::universal_memory_resource>;
+
+} // namespace
+
+template <typename T>
+void TestUniversalPointer()
+{
+  // Simple test to ensure that pointers created with universal_memory_resource
+  // can be dereferenced and used with STL code. This is necessary as some
+  // STL implementations break when using fancy references that overload
+  // `operator&`, so universal_memory_resource uses a special pointer type that
+  // returns regular C++ references that can be safely used host-side.
+
+  // These operations fail to compile with fancy references:
+  auto p0 = thrust::allocate_unique<T>(universal_allocator<T>{});
+  p0.get();
+  *p0;
+  std::vector<T, universal_allocator<T>> v0;
```

Comment:
Is this enough? Should we try instantiating some of the more common member functions that actually try to invoke operations on the value and pointer types, and not just the default constructor?

Reply:
These were enough to expose the compile failures with `cuda::pointer`, so it's enough in that sense. If there's more API that you think should be tested here I'm happy to add more.

Reply:
I think a better way to test this would be to use a `std::vector` with the universal allocator as part of the vector unit test macros, that would give much better coverage of this.

I'll get that working and update this PR when it's done.

Reply:
I've added some basic validation of using a `std::vector<T, universal_allocator>` with STL algorithms here instead.